RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernelsImpl.h

```diff
@@ -0,0 +1,593 @@
+    }
+  }
+}
+
+__global__ void kernel_mark_used(GPUCACell::Hits const *__restrict__ hhp,
+                                 GPUCACell *__restrict__ cells,
+                                 uint32_t const *nCells) {
+  auto first = threadIdx.x + blockIdx.x * blockDim.x;
+  for (int idx = first, nt = (*nCells); idx < nt; idx += gridDim.x * blockDim.x) {
+    auto &thisCell = cells[idx];
+    if (!thisCell.tracks().empty())
+      thisCell.setUsedBit(2);
+  }
+}
+
+__global__ void kernel_countMultiplicity(HitContainer const *__restrict__ foundNtuplets,
+                                         Quality const *__restrict__ quality,
+                                         caConstants::TupleMultiplicity *tupleMultiplicity) {
+  auto first = blockIdx.x * blockDim.x + threadIdx.x;
+  for (int it = first, nt = foundNtuplets->nbins(); it < nt; it += gridDim.x * blockDim.x) {
+    auto nhits = foundNtuplets->size(it);
+    if (nhits < 3)
+      continue;
+    if (quality[it] == pixelTrack::Quality::dup)
+      continue;
+    assert(quality[it] == pixelTrack::Quality::bad);
+    if (nhits > 5)
+      printf("wrong mult %d %d\n", it, nhits);
+    assert(nhits < 8);
+    tupleMultiplicity->countDirect(nhits);
+  }
+}
+
+__global__ void kernel_fillMultiplicity(HitContainer const *__restrict__ foundNtuplets,
+                                        Quality const *__restrict__ quality,
+                                        caConstants::TupleMultiplicity *tupleMultiplicity) {
+  auto first = blockIdx.x * blockDim.x + threadIdx.x;
+  for (int it = first, nt = foundNtuplets->nbins(); it < nt; it += gridDim.x * blockDim.x) {
+    auto nhits = foundNtuplets->size(it);
+    if (nhits < 3)
+      continue;
+    if (quality[it] == pixelTrack::Quality::dup)
+      continue;
+    assert(quality[it] == pixelTrack::Quality::bad);
+    if (nhits > 5)
+      printf("wrong mult %d %d\n", it, nhits);
+    assert(nhits < 8);
+    tupleMultiplicity->fillDirect(nhits, it);
+  }
+}
+
+__global__ void kernel_classifyTracks(HitContainer const *__restrict__ tuples,
+                                      TkSoA const *__restrict__ tracks,
+                                      CAHitNtupletGeneratorKernelsGPU::QualityCuts cuts,
+                                      Quality *__restrict__ quality) {
+  int first = blockDim.x * blockIdx.x + threadIdx.x;
+  for (int it = first, nt = tuples->nbins(); it < nt; it += gridDim.x * blockDim.x) {
+    auto nhits = tuples->size(it);
+    if (nhits == 0)
+      break;  // guard
+
+    // if duplicate: not even fit
+    if (quality[it] == pixelTrack::Quality::dup)
+      continue;
+
+    assert(quality[it] == pixelTrack::Quality::bad);
+
+    // mark doublets as bad
+    if (nhits < 3)
+      continue;
+
+    // if the fit has any invalid parameters, mark it as bad
+    bool isNaN = false;
+    for (int i = 0; i < 5; ++i) {
+      isNaN |= std::isnan(tracks->stateAtBS.state(it)(i));
```

Comment:
isNotFinite has not been tested on GPU (at best of my knowledge).
Please open a issue on the subject

Reply:
It is not shocking at first sight (details in the issue: https://github.com/cms-patatrack/cmssw/issues/617).

Reply:
I admit I'm not familiar with the history here.
@VinInn @slava77 , why is `edm::isNotFinite()` preferred over `std::isnan()` or `not std::isfinite()` ?

Reply:
> 
> I admit I'm not familiar with the history here. @VinInn @slava77 , why is edm::isNotFinite() preferred over std::isnan() or not std::isfinite() ?
> because if you compile with -Ofast finite-math is assumed and the compiler just remove the test all together... v.

I would expect the intrinsic to be honored by gcc and to generate something based on the instruction `FXAM—Examine Floating-Point`, while -Ofast would mean masking the floating point exceptions in the FPU flags with `FSTCW/FNSTCW—Store x87 FPU Control Word`. Speculations to be verified...

Reply:
I see, thanks for the explanation.

A quick check with godbolt says that `std::isfinite()` and `__builtin_isfnite()` are just compiled out with `-ffast-math` / `-Ofast`.

### source code
```c++
#include <cmath>

bool std_is_finite(float value) {
    return std::isfinite(value);
}

bool gcc_is_finite(float value) {
    return __builtin_isfinite(value);
}

bool edm_is_finite(float value) {
    constexpr unsigned int mask = 0x7f800000;
    union {
      unsigned int l;
      float d;
    } v = {.d = value};
    return (v.l & mask) != mask;
}
```

### compiled with GCC 10.3 with `-std=c++17 -O2 -march=haswell`
```asm
std_is_finite(float):
        vandps  xmm0, xmm0, XMMWORD PTR .LC0[rip]
        vmovss  xmm1, DWORD PTR .LC1[rip]
        vucomiss        xmm1, xmm0
        setnb   al
        ret
gcc_is_finite(float):
        vandps  xmm0, xmm0, XMMWORD PTR .LC0[rip]
        vmovss  xmm1, DWORD PTR .LC1[rip]
        vucomiss        xmm1, xmm0
        setnb   al
        ret
edm_is_finite(float):
        vmovd   DWORD PTR [rsp-4], xmm0
        mov     eax, 2139095040
        and     eax, DWORD PTR [rsp-4]
        cmp     eax, 2139095040
        setne   al
        ret
.LC0:
        .long   2147483647
        .long   0
        .long   0
        .long   0
.LC1:
        .long   2139095039
```

### compiled with GCC 10.3 with `-std=c++17 -Ofast -march=haswell`
```asm
std_is_finite(float):
        mov     eax, 1
        ret
gcc_is_finite(float):
        mov     eax, 1
        ret
edm_is_finite(float):
        vmovd   DWORD PTR [rsp-4], xmm0
        mov     eax, 2139095040
        and     eax, DWORD PTR [rsp-4]
        cmp     eax, 2139095040
        setne   al
        ret
```

Reply:
While with CUDA, all of `std::isfinite()`, `__builtin_isfinite()`, and `::isfinite()` are kept at `-O3 --use_fast_math`.

### source code
```cuda
#include <cmath>

__global__
void std_is_finite(float value, bool* result) {
    *result = std::isfinite(value);
}

__global__
void gcc_is_finite(float value, bool* result) {
    *result = __builtin_isfinite(value);
}

__global__
void cuda_is_finite(float value, bool* result) {
    *result = ::isfinite(value);
}

__global__
void edm_is_finite(float value, bool* result) {
    constexpr unsigned int mask = 0x7f800000;
    union {
      unsigned int l;
      float d;
    } v = {.d = value};
    *result = (v.l & mask) != mask;
}
```

### compiled with NVCC 11.2 with `-std=c++17 -O3 --use_fast_math -gencode arch=compute_70,code=sm_70`
```asm
.visible .entry _Z13std_is_finitefPb(
        .param .f32 _Z13std_is_finitefPb_param_0,
        .param .u64 _Z13std_is_finitefPb_param_1
)
{

        ld.param.f32    %f1, [_Z13std_is_finitefPb_param_0];
        ld.param.u64    %rd1, [_Z13std_is_finitefPb_param_1];
        cvta.to.global.u64      %rd2, %rd1;
        abs.ftz.f32     %f2, %f1;
        setp.lt.ftz.f32         %p1, %f2, 0f7F800000;
        selp.u16        %rs1, 1, 0, %p1;
        st.global.u8    [%rd2], %rs1;
        ret;

}
.visible .entry _Z13gcc_is_finitefPb(
        .param .f32 _Z13gcc_is_finitefPb_param_0,
        .param .u64 _Z13gcc_is_finitefPb_param_1
)
{

        ld.param.f32    %f1, [_Z13gcc_is_finitefPb_param_0];
        ld.param.u64    %rd1, [_Z13gcc_is_finitefPb_param_1];
        cvta.to.global.u64      %rd2, %rd1;
        abs.ftz.f32     %f2, %f1;
        setp.lt.ftz.f32         %p1, %f2, 0f7F800000;
        selp.u16        %rs1, 1, 0, %p1;
        st.global.u8    [%rd2], %rs1;
        ret;

}
.visible .entry _Z14cuda_is_finitefPb(
        .param .f32 _Z14cuda_is_finitefPb_param_0,
        .param .u64 _Z14cuda_is_finitefPb_param_1
)
{

        ld.param.f32    %f1, [_Z14cuda_is_finitefPb_param_0];
        ld.param.u64    %rd1, [_Z14cuda_is_finitefPb_param_1];
        cvta.to.global.u64      %rd2, %rd1;
        abs.ftz.f32     %f2, %f1;
        setp.lt.ftz.f32         %p1, %f2, 0f7F800000;
        selp.u16        %rs1, 1, 0, %p1;
        st.global.u8    [%rd2], %rs1;
        ret;

}
.visible .entry _Z13edm_is_finitefPb(
        .param .f32 _Z13edm_is_finitefPb_param_0,
        .param .u64 _Z13edm_is_finitefPb_param_1
)
{

        ld.param.u64    %rd1, [_Z13edm_is_finitefPb_param_1];
        cvta.to.global.u64      %rd2, %rd1;
        ld.param.u32    %r1, [_Z13edm_is_finitefPb_param_0];
        and.b32         %r2, %r1, 2139095040;
        setp.ne.s32     %p1, %r2, 2139095040;
        selp.u16        %rs1, 1, 0, %p1;
        st.global.u8    [%rd2], %rs1;
        ret;

}
```

Reply:
So, in device code `isfnite()` & co should be fine.

Reply:
> I admit I'm not familiar with the history here.
> @VinInn @slava77 , why is `edm::isNotFinite()` preferred over `std::isnan()` or `not std::isfinite()` ?

I was echoing the static analyzer category

fastmath plugin |  
-- | --
std::isnan / std::isinf does not work when fast-math is used.  Please use edm::isNotFinite from 'FWCore/Utilities/interface/isFinite.h'


this file is not reported in the analyzer; so perhaps it's not an issue after all.

Reply:
On the other hand the `plugins` appears to be compiled with `-Ofast`
https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoPixelVertexing/PixelTriplets/plugins/BuildFile.xml#L2
and this header (along this function) is compiled for host via
https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernels.cc#L162

Reply:
I will add the edm function, which is very close to what CUDA does.

I wrongly assumed we get floating point exceptions by default. Quite the opposite we need to explicitly require them:
```c++
#include <fenv.h>

feenableexcept (FE_DIVBYZERO|FE_...);
```